becquerel/core/fitting.py

```diff
@@ -0,0 +1,681 @@
+# -----------------------------------------------------------------------------
+
+
+# Helper functions for guessing
+def _xy_right(y, x=None, num=4):
+    if x is not None:
+        return np.mean(x[-num:]), np.mean(y[-num:])
+    return len(y)*0.5, np.mean(y[-num:])
+
+
+def _xy_left(y, x=None, num=4):
+    if x is not None:
+        return np.mean(x[:num]), np.mean(y[:num])
+    return len(y)*0.5, np.mean(y[:num])
+
+
+class ConstantModel(Model):
+
+    def __init__(self, *args, **kwargs):
+        super(ConstantModel, self).__init__(constant, *args, **kwargs)
+        # TODO: remove this min setting?
+        self.set_param_hint('{}c'.format(self.prefix), min=0.)
+
+    def guess(self, y, x=None, num=2):
+        return []
+
+
+class LineModel(Model):
+
+    def __init__(self, *args, **kwargs):
+        super(LineModel, self).__init__(line, *args, **kwargs)
+
+    def guess(self, y, x=None, num=2):
+        _, b = _xy_left(y, x=x, num=num)
+        return [
+            ('{}m'.format(self.prefix), 'value', 0.),
+            ('{}b'.format(self.prefix), 'value', b),
+        ]
+
+
+class GaussModel(Model):
+
+    def __init__(self, *args, **kwargs):
+        super(GaussModel, self).__init__(gauss, *args, **kwargs)
+        self.set_param_hint(
+            '{}fwhm'.format(self.prefix),
+            expr='{} * {}sigma'.format(FWHM_SIG_RATIO, self.prefix))
+
+    def guess(self, y, x=None, center_ratio=0.5, width_ratio=0.5):
+        assert center_ratio < 1, \
+            'Center mask ratio cannot exceed 1: {}'.format(center_ratio)
+        assert width_ratio < 1, \
+            'Width mask ratio cannot exceed 1: {}'.format(width_ratio)
+        if x is None:
+            x = np.arange(0, len(y))
+        xspan = x[-1] - x[0]
+        mu = x[0] + xspan * center_ratio
+        msk = ((x >= (mu - xspan * width_ratio)) &
+               (x <= mu + xspan * width_ratio))
+        # NOTE: this integration assumes y is NOT normalized to dx
+        amp = np.sum(y[msk])
+        # TODO: update this, minimizer creates NaN's if default sigma used (0)
+        sigma = xspan * width_ratio / 10.
+        return [
+            ('{}amp'.format(self.prefix), 'value', amp),
+            ('{}amp'.format(self.prefix), 'min', 0.0),
+            ('{}mu'.format(self.prefix), 'value', mu),
+            ('{}mu'.format(self.prefix), 'min', x[0]),
+            ('{}mu'.format(self.prefix), 'max', x[-1]),
+            ('{}sigma'.format(self.prefix), 'value', sigma),
+            ('{}sigma'.format(self.prefix), 'min', 0.0),
+        ]
+
+
+class ErfModel(Model):
+
+    def __init__(self, *args, **kwargs):
+        super(ErfModel, self).__init__(erf, *args, **kwargs)
+
+    def guess(self, y, x=None):
+        return [
+            ('{}amp'.format(self.prefix), 'value', y[0] - y[-1]),
+            ('{}amp'.format(self.prefix), 'min', 0.),
+            ('{}mu'.format(self.prefix), 'expr', 'gauss_mu'),
+            ('{}sigma'.format(self.prefix), 'expr', 'gauss_sigma'),
```

Comment:
Is there a way to generalize this? I.e. what if there are two Gaussians, or if we like to fit ExpGauss + Erf?

Reply:
the names are generalized. are you talking about the guess values?

Reply:
Sorry, wasn't too precise here. I talk about `erf_mu` being intialized as `gauss_mu` and the same for `sigma`. If you have a ExpGauss + Erf model, there is no parameter `gauss_mu`, so a call like `Fitter(['expgauss', 'erf'])` or simply `Fitter(['erf'])` fails.

I don't think we can make this more general, I think however we should be careful when we define parameters (using `make_params`, or `set_params`), so that the user can fix a conflict before something fails.